gateway/sensor.go

```diff
@@ -174,139 +232,85 @@ func (g *Gateway) receivePackets() {
 			default:
 			}
 			numPackets := int(C.receive(packet))
-			if numPackets < 0 || numPackets > 1 {
-				g.logger.Errorf("error receiving lora packet")
-			}
-			if numPackets != 0 {
+			switch numPackets {
+			case 0:
+				// no packet received, wait 10 ms to receive again.
+				select {
+				case <-ctx.Done():
+					return
+				case <-time.After(10 * time.Millisecond):
+				}
+			case 1:
+				// received a LORA packet
 				var payload []byte
 				for i := 0; i < numPackets; i++ {
+					if packet.size == 0 {
+						continue
+					}
 					// Convert packet to go byte array
 					for i := 0; i < int(packet.size); i++ {
 						payload = append(payload, byte(packet.payload[i]))
 					}
-					g.handlePacket(payload)
-				}
-			} else {
-				// wait 10 ms to receive another packet
-				select {
-				case <-ctx.Done():
-					return
-				case <-time.After(10 * time.Millisecond):
+					g.handlePacket(ctx, payload)
 				}
+			default:
+				g.logger.Errorf("error receiving lora packet")
 			}
 		}
 	})
 }
 
-func (g *Gateway) handlePacket(payload []byte) {
-	// first byte is MHDR - specifies message type
-	switch payload[0] {
-	case 0x0:
-		g.logger.Errorf("OTAA not supported - use ABP")
-	case 0x40:
-		g.logger.Infof("received data uplink")
-		name, readings, err := g.parseDataUplink(payload)
-		if err != nil {
-			g.logger.Errorf("error parsing uplink message: %w", err)
-		} else if name != "" {
-			// the milesight lora sensor sends two data uplink messages close together, we want both of these in the map.
+func (g *Gateway) handlePacket(ctx context.Context, payload []byte) {
+	g.workers.Add(func(ctx context.Context) {
+		// first byte is MHDR - specifies message type
+		switch payload[0] {
+		case 0x0:
+			g.logger.Infof("recieved join request")
```

Comment:
Should these be debug logs? Also fine if you wanna leave while the project is in progress

Reply:
yeah I can change them to debug later on